presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java

```diff
@@ -38,42 +48,92 @@
 {
     private final Supplier<ExternalPlanStatisticsProvider> externalPlanStatisticsProvider;
     private final StatsCalculator delegate;
+    private final ObjectMapper objectMapper;
 
-    public HistoryBasedPlanStatisticsCalculator(Supplier<ExternalPlanStatisticsProvider> externalPlanStatisticsProvider, StatsCalculator delegate)
+    // TODO: Evict plan node hashes once a query finishes.
+    private final ConcurrentMap<QueryId, CanonicalPlanHashes> planNodeHashes;
+
+    public HistoryBasedPlanStatisticsCalculator(
+            Supplier<ExternalPlanStatisticsProvider> externalPlanStatisticsProvider,
+            StatsCalculator delegate,
+            ObjectMapper objectMapper)
     {
         this.externalPlanStatisticsProvider = requireNonNull(externalPlanStatisticsProvider, "externalPlanStatisticsProvider is null");
         this.delegate = requireNonNull(delegate, "delegate is null");
+        this.objectMapper = requireNonNull(objectMapper, "objectMapper is null");
+        this.planNodeHashes = new ConcurrentHashMap<>();
     }
 
     @Override
     public PlanNodeStatsEstimate calculateStats(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, TypeProvider types)
     {
         return delegate.calculateStats(node, sourceStats, lookup, session, types)
-                .combineStats(getStatistics(node, session, types, lookup));
+                .combineStats(getStatistics(node, session, lookup));
     }
 
-    private PlanNode removeGroupReferences(PlanNode planNode, Lookup lookup)
+    public void cacheHashBasedStatistics(Session session, PlanNode root, TypeProvider types)
     {
-        if (planNode instanceof GroupReference) {
-            return removeGroupReferences(lookup.resolve(planNode), lookup);
+        if (!useExternalPlanStatisticsEnabled(session)) {
+            return;
+        }
+        CanonicalPlanHashes canonicalPlanHashes = new CanonicalPlanHashes(root, objectMapper);
+        planNodeHashes.put(session.getQueryId(), canonicalPlanHashes);
+
+        ImmutableList.Builder<PlanNodeWithHash> hashes = new ImmutableList.Builder<>();
+        Stack<PlanNode> stack = new Stack<>();
+        stack.push(root);
+        while (!stack.isEmpty()) {
+            PlanNode plan = stack.pop();
+            plan.getSources().forEach(stack::push);
+            for (String hash : canonicalPlanHashes.getCanonicalPlanHashes(plan.getId()).values()) {
+                hashes.add(new PlanNodeWithHash(plan, Optional.of(hash)));
+            }
         }
-        List<PlanNode> children = planNode.getSources().stream().map(node -> removeGroupReferences(node, lookup)).collect(toImmutableList());
-        return planNode.replaceChildren(children);
+        // Cache statistics for future calls
+        // TODO: Right now, we assume that underlying implementation will cache statistics
```

Comment:
this to do is pretty significant because it breaks the whole abstraction. The point of interfaces is  to remove any assumptions about how the underlying implementations work.  At the very least, this should be calling a method called something like "cachePlanStats".  That's still not ideal because you're telling the implementation how to do it's job, but at least it's explicit.  But I bet there's a cleaner way to structure the code so that we're not asking for caching across the interface boundary (e.g. we never need to ask for things to be cached, and implementations just do it as they see fit. or alternatively, we have the cache on this side).

Reply:
Skipped caching for now as PR is getting too big, just storing hashes.